### PR TITLE
UI extraction performance improvements

### PR DESCRIPTION
**Pull request by ickshonpe**
# Objective

UI extraction is really inefficient:
* Each extraction function has to walk the entire UiStack and call `Query::get` for each entity. So if there are 10000 nodes it will do 10000 lookups, even if there are no entities that match the query.
* `prepare_uinodes` does an in-place stable sort of the `ExtractedUiNodes::uinodes` vector. It's extremely expensive as each `ExtractedUiNode` is quite large (192 bytes) and there can be a lot of them.

## Solution

Add a new component `UiStackIndex` to hold each UI node's stack index generated by `ui_stack_system`. Then during extraction we can query for each node's stack index and we no longer have to iterate the UiStack each time.

`ExtractedUiNodes` now has a second vec of
```
struct ExtractedRange {
    stack_index: u32,
    range: Range<u32>,
}
```
These ranges partition the `ExtractedUinodes::uinodes` vector.

Now instead of sorting `uinodes` we can sort the `ExtractedRange`s vector by stack index instead. This is faster because  `ExtractedRange`s are smaller and there are fewer of them (a page of text using thousands of `ExtractedUiNode`s only requires one `ExtractedRange`).

### Benchmarks

```
cargo run --profile stress-test --features trace_tracy --example many_buttons
```
prepare

text

Capture

atlas

### Notes

* The PR is quite small. The large number of lines changed is due to the elimination of some if-let blocks reducing indentation, most of the logic is unaltered.

## Changelog
* `ExtractedUiNodes::uinodes` is no longer sorted in `prepare_uinodes`, instead it has a second vec of ranges that partition `uinodes`. This partition is sorted by stack index instead.
* New component `UiStackIndex`: holds each UI node's stack index as generated by `ui_stack_system`.
* UI extraction functions no longer walk `UiStack`, instead the stack index for each node is retrieved from their `UiStackIndex` component.
* The `uinodes` field of `ExtractedUiNodes` is now private. Use the new `push` and `push_nodes` to add `ExtractedUiNode`s to be rendered.

## Migration Guide
The `uinodes` field of `ExtractedUiNodes` is now private. Use the new `push` and `push_nodes` to add `ExtractedUiNode`s to be rendered.

### CODE REVIEW

File: crates/bevy_ui/src/render/mod.rs
Comment (nicopap) [inline crates/bevy_ui/src/render/mod.rs:692]: As you mentioned in the PR description, this needs to be a method on `ExtractedUiNode`. Ideally an `impl Iterator` `next` method.

File: crates/bevy_ui/src/render/mod.rs
Comment (nicopap) [inline crates/bevy_ui/src/render/mod.rs:204]: Nit: I like short names, but since it's only used once, could we get a more descriptive name here?

```suggestion
    let nodes_buffer = extracted_uinodes.next_buffer();
```

File: crates/bevy_ui/src/render/mod.rs
Comment (nicopap) [inline crates/bevy_ui/src/render/mod.rs:167]: Since the number of extract functions is already known, could this be replaced by a `[Vec<ExtractedUiNode>; 4]`?

File: crates/bevy_ui/src/render/mod.rs
Comment (nicopap) [inline crates/bevy_ui/src/render/mod.rs:689]: I was at first extremely confused by how this works. So here is a summary for other reviewers:

1. find the index (named `n`) in `drains` where `node.stack_index` is smallest (the `for` loop)
2. Pick the `(current, drain): (Option<T>, Drain<T>)` from `drains`
3. Read the `current`
4. replace it with the `drain.next()`
5. return `current`

I might suggest some improvement in a couple hours.

**Review by nicopap — COMMENTED** [review]
The idea is sound. Really nice to take advantage of the already-sorted nature of the UiStack though I think the code quality could be improved.

**Comment by nicopap** [comment]
I opened two PRs to your branch, please check them out:

1. Using unsafe https://github.com/ickshonpe/bevy/pull/2
2. No unsafe, but more allocation https://github.com/ickshonpe/bevy/pull/1

File: crates/bevy_ui/src/render/mod.rs
Comment (ickshonpe) [reply crates/bevy_ui/src/render/mod.rs:167]: Needs to be a Vec I think so users still add their own extraction functions to the schedule.

File: crates/bevy_ui/src/render/mod.rs
Comment (ickshonpe) [reply crates/bevy_ui/src/render/mod.rs:689]: Yes it's weird I just typed this in and it worked immediately, and then all the better-thought-out refactors I tried ruined the performance.

**Comment by nicopap** [comment]
An alternative implementation is to create a `Vec<ExtractedUiNode>`, (maybe using a `SparseSet`?) pass it as a resource to all the extract functions, and insert the relevant node at the index in UiStack, now we have an already-sorted `Vec`!

The downside is that it might be fairly wasteful if _most_ ui nodes are invisible (supposing we simply "hide" menus instead of despawning them) The current impl might consume at most twice as much memory as there are visible extracted UI nodes though the memory for extracted UI nodes is never reclaimed.

File: crates/bevy_ui/src/render/mod.rs
Comment (nicopap) [reply crates/bevy_ui/src/render/mod.rs:167]: A yeah, then it can't be optimized this way. I see already a few crates relying on `ExtractedUiNodes` https://github.com/search?q=ExtractedUiNodes&type=code

**Comment by nicopap** [comment]
I traced my PRs and they are a bit slower than your code.

**Comment by ickshonpe** [comment]
> I traced my PRs and they are a bit slower than your code.

Yeah I can't figure it out. I came back a couple of times yesterday to try a few new ideas but either they made no difference or were significantly worse (those were ones I thought were really promising).

Reimplemented the original version as an iterator and made a few refactorings so it's not so ugly anyway. No performance difference:

Captureprep

**Comment by nicopap** [comment]
I did some more experimenting and I found we should also check the time on `ExtractSchedule`, since inserting into `ExtractedUiNodes` takes time as well.

**Comment by nicopap** [comment]
Using a sparse set (`Vec<Option<ExtractedUiNode>>`) is much faster than a Illife vector (`Vec<Vec<ExtractedUiNode>>`) I opened a PR on your branch, but with caveats:

* https://github.com/ickshonpe/bevy/pull/3

**Comment by nicopap** [comment]
Further design discussion is available in this discord conversation: https://discord.com/channels/691052431525675048/743663673393938453/1134046753692336249

Basically currently my sparse set proposal is incompatible with how bevy handles text, since it uploads each glyph as an individual "extracted UI node"

File: crates/bevy_ui/src/ui_node.rs
Comment (nicopap) [inline crates/bevy_ui/src/ui_node.rs:82]: Does it make sense to make `.0` pub? Users might expect that setting it does something but it's set unconditionally in `ui_stack_system`.

**Review by nicopap — APPROVED** [review]
Excellent solution. Code quality is high.

Basically adds a list of range to `ExtractedUiNodes` and sorts the spans instead of the nodes themselves. This is useful because:

* text nodes have a single extracted node per **glyph**, so we could easily see hundred of nodes
* nodes with border spawn actually five extracted nodes

So we may divide by an order of magnitude the number of sorted items.

Though I think the need for this optimization stems mostly from a shortcoming in the way we handle ui nodes. It's just weird to have an "extracted" version that doesn't correspond to the actual version. Feels like bad indirection. But that's for another time.

File: crates/bevy_ui/src/ui_node.rs
Comment (ickshonpe) [reply crates/bevy_ui/src/ui_node.rs:82]: Yep, this was meant to be private with an accessor function, I'll change it.

File: crates/bevy_ui/src/node_bundles.rs
Comment (ickshonpe) [reply crates/bevy_ui/src/node_bundles.rs:123]: I copied from the doc comments for the other fields without thinking. Yeah, it doesn't make sense. The bundle's field isn't the thing being managed by the system. Better change them all.

Comment (nicopap) [reply crates/bevy_ui/src/node_bundles.rs:123]: If you prefer, you could open an issue and keep this for a different PR.

Comment (ickshonpe) [reply crates/bevy_ui/src/node_bundles.rs]: Just noticed as well for `GlobalTransform`:
```
/// This component is automatically managed by the UI layout system.
```
I think I'll fix the `field` thing here as it's within the focus of this PR as it affects the new comments I added, but fix the GlobalTransform comment in another PR.

**Comment by mockersf** [comment]
I relaunched CI, tests are still failing, could you take a look?

File: crates/bevy_ui/src/node_bundles.rs
Comment (alice-i-cecile) [inline crates/bevy_ui/src/node_bundles.rs:56]: Link would be nice here.

File: crates/bevy_ui/src/node_bundles.rs
Comment (alice-i-cecile) [inline crates/bevy_ui/src/node_bundles.rs:39]: Links would be helpful here and below.

File: crates/bevy_ui/src/node_bundles.rs
Comment (alice-i-cecile) [reply crates/bevy_ui/src/node_bundles.rs:56]: Ditto replicated doc comments before.

**Review by alice-i-cecile — APPROVED** [review]
Code and docs quality is good, motivation is sound and benchmark results are excellent.

File: crates/bevy_ui/src/node_bundles.rs
Comment (ickshonpe) [reply crates/bevy_ui/src/node_bundles.rs:56]: Ah, can't link to `ui_stack_system` because it's private, that's why the CI was failing.

**Comment by ickshonpe** [comment]
> I relaunched CI, tests are still failing, could you take a look?

should be good now

**Comment by alice-i-cecile** [comment]
>  stack::tests::test_ui_stack_system

Test still appears to be failing 🤔

**Comment by ickshonpe** [comment]
> > stack::tests::test_ui_stack_system
> 
> Test still appears to be failing 🤔

I'll have another look

**Comment by alice-i-cecile** [comment]
@ickshonpe I can merge this once the merge conflict is resolved :) Ping me if I'm slow!

**Comment by ickshonpe** [comment]
> @ickshonpe I can merge this once the merge conflict is resolved :) Ping me if I'm slow!

I might open a new PR and leave this one. I've had to rework everything to accommodate the recent rendering changes.

**Comment by alice-i-cecile** [comment]
Closing in favor of #9668 due to merge conflicts.

**Pull request closed without merging**
